scientific-computing-with-python/learn-encapsulation-by-drawing-a-projectile-trajectory/projectile-trajectory.py

```diff
@@ -0,0 +1,122 @@
+import math
+import re
+
+# constants
+g = 9.81
+projectile = '∙'
+
+
+class Projectile:
+
+    def __init__(self, height, speed, angle):
+        self.__height = height
+        self.__speed = speed
+        self.__angle = math.radians(angle)
+
+        self.__calculate_displacement()
+        self.__calculate_all_coordinates()
+
+    def __calculate_displacement(self):
+        self.__displacement = self.__speed * math.cos(self.__angle) * (
+            self.__speed * math.sin(self.__angle) +
+            math.sqrt(self.__speed**2 * math.sin(self.__angle)**2 +
+                      2 * g * self.__height)) / g
+
+    def __calculate_y_coordinate(self, x):
+        return self.__height + math.tan(self.__angle) * x - g * x**2 / (
+            2 * self.__speed**2 * math.cos(self.__angle)**2)
+
+    def __calculate_all_coordinates(self):
+        self.__coordinates = []
+        for x in range(math.ceil(self.__displacement)):
+            y = self.__calculate_y_coordinate(x)
+            self.__coordinates.append((x, y))
+
+    def __create_coordinates_table(self):
+        table = '  x      y\n'
+        for x, y in self.__coordinates:
+            table += f'{x:>3}{round(y, 2):>7.2f}\n'
+
+        return table
+
+    def __create_trajectory(self):
+        rounded_coord = [(x, round(y)) for x, y in self.__coordinates]
+        rows = max([y for x, y in rounded_coord]) + 1
+        columns = max([x for x, y in rounded_coord]) + 1
+        # TODO?: if rows and/or columns too big, no draw (how big?)
+        graph = ''
+        for y in range(rows, -3, -1):
+            row = ''
+            for x in range(-2, columns + 1):
+                if (x, y) in rounded_coord:
+                    row += projectile
+                elif x == -1 and y >= 0:
+                    row += '⊣'
+                elif y == -1 and x >= 0:
+                    row += '⊤'
+                elif (x == -2 and y == 0) or (x == 0 and y == -2):
+                    row += '0'
+                else:
+                    row += ' '
+            graph += row + '\n'
```

Comment:
It's hard to understand what is changed when. I'd separate composing axes representations from projectile graph. Instead of starting with string for a whole string representation, `graph` could be 2d list, with `rows` x `columns`. First insert axes, then handle projectile points, in final step join everything to string.

Reply:
I have made a different drawing function